ananse/enhancer_binding.py

```diff
@@ -0,0 +1,497 @@
+        fix_outliers=False,
+        output_bed3=True,
+        verbose=True,
+    ):
+        """
+        Set bed region width.
+
+        If the input bed is a narrowPeak file (narrowpeak=True),
+        center region on the summit (start+peak).
+        Otherwise center on the middle of the region.
+
+        If fix_outliers is set to True, shift regions to fit their chromosomes.
+        Otherwise drop these regions.
+
+        If output_bed3 is set to False, output the whole bed file.
+        """
+        half_seqlen = width // 2
+        chrom_sizes = genomepy.Genome(genome).sizes
+        missing_chrm = []
+
+        if narrowpeak:
+
+            def get_summit(_start, _, summit_offset):
+                return _start + int(summit_offset)
+
+            summit_col = 9
+        else:
+
+            def get_summit(_start, _end, _):
+                return (_start + _end) // 2
+
+            summit_col = 0  # unused
+
+        with open(bed_in) as old, open(bed_out, "w") as new:
+            for line in old:
+                if line.startswith("#"):
+                    continue
+                line = line.split("\t")
+                chrm = str(line[0])
+                if chrm not in chrom_sizes.keys():
+                    missing_chrm.append(chrm)
+                    continue
+                start = int(line[1])
+                end = int(line[2])
+                rest = line[3:] if not output_bed3 else []
+
+                chrm_len = chrom_sizes[chrm]
+                if width == end - start:
+                    nstart = str(start)
+                    nend = str(end)
+                elif chrm_len <= width:
+                    if not fix_outliers:
+                        continue
+                    nstart = str(0)
+                    nend = str(chrm_len)
+                else:
+                    summit = get_summit(start, end, line[summit_col])
+                    if not fix_outliers:
+                        nstart = str(summit - half_seqlen)
+                        nend = str(summit + half_seqlen)
+                        if int(nstart) < 0 or int(nend) > chrm_len:
+                            continue
+                    else:
+                        # adjust the summit for the chromosome boundaries
+                        summit = max(summit, 0 + half_seqlen)
+                        summit = min(summit, chrm_len - half_seqlen)
+
+                        nstart = str(summit - half_seqlen)
+                        nend = str(summit + half_seqlen)
+
+                new.write("\t".join([chrm, nstart, nend] + rest) + "\n")
+
+        if missing_chrm and verbose:
+            print(
+                "The following contigs were present in "
+                + f"'{os.path.basename(bed_in)}', "
+                + "but were missing in the genome file: "
+                + f"{', '.join(list(set(missing_chrm)))}\n"
+            )
+        return bed_out
+
+    def run(self, outfile, width=200, force=False):
```

Comment:
why not gimmemotifs.preprocessing.combine_peaks?

Reply:
Quan's original input were broadPeak files (no summit columns). 

My implementation can handle any BED format files, and will utilize the narrowPeak's summit values if available.

Reply:
What do you mean by using the summit values? What combine_peaks does is also looking for the highest summit values, and splitting peaks if necessary, some extra logic. I would be in favor of implementing extra functionality there if needed, so that we don't have to implement the same type of function in two different places :)

Reply:
> don't have to implement the same type of function in two different places :)

If I had know gimme had a similar function I would have tried to expand that, oops!

> using the summit values

if the input is narrowPeak, the resized region is centered on the `start + summit`, else it is centered on the `end-start // 2`

> implementing extra functionality

`combine_peaks` can currently only read narrowPeak and summits.bed files. Since Quan used broadPeak files, I figures we need to be able to read any BED file. Jos and I discussed that BED3 would also be a great format as input, so that confirms that.

If `combine_peaks` can be expanded to read any BED format file, it would better (since it can also use summits from summit.bed files)